### PR TITLE
nixos: Provide users.setupHomeScript option

### PR DESCRIPTION
**Pull request by eskimor**
Provide a new option which allows the user to fill a user's home
directory with initial content.

Tested it thoroughly on my system - Worked like a charm :-)

The perl code can possibly be improved - I never wrote any perl before.

### CODE REVIEW

File: nixos/modules/config/users-groups.nix [hunk 1]
```diff
@@ -446,6 +446,26 @@ in {
       options = [ groupOpts ];
     };
 
+
+    users.setupHomeScript = mkOption {
+      type = types.string;
+      default = "";
+      description = ''
+        The specified script will be called when the home directory of the user
+        gets created. It will not be called if the directory already exists.
```
Comment (jagajaga) [inline nixos/modules/config/users-groups.nix]: If it called once, maybe rename it to smth like `initialSetup`?

**Comment by jagajaga** [comment]
cc @edolstra

**Comment by eskimor** [comment]
initialHomeSetup instead of setupHomeScript ? Or include the "Script" part: initialHomeSetupScript is getting lengthy though. Opinions?

**Comment by eskimor** [comment]
Ok, I settled with initialHomeSetup and made a little improvement: The implementation now takes care that the created files have the right owner and group.

**Comment by eskimor** [comment]
Any more feedback? I am very willing to improve this request, if concerns are raised! :-)

**Comment by jagajaga** [comment]
Well, this feature is nice, but I think it's a little bit... Well, for example. If I used it once and then I realized that I was wrong -> next `nixos-rebuild` will call this `init`? Maybe you can go deeper and implement some kind of user's dot files manager via `users.<user>.dotFiles = [list of files]`, or even use the same technique as we use in `environment.etc."file name"`?

**Comment by edolstra** [comment]
Yeah, I agree with @jagajaga. This is more imperative than declarative, especially since the setup script is only run when the home directory is first created. A more declarative behavior would be something like `environment.etc`.

For specifying the initial contents of a home directory, the PAM module has an option `makeHomeDir` that copies the contents of `/etc/skel` on initial login, but I haven't tried it myself.

**Comment by eskimor** [comment]
Thank you for the feedback! I'll look into environment.etc.

**Comment by nbp** [comment]
Honestly, I do not see any issue for having an imperative way of setting the content, and having a declarative method which use the imperative one.

We could make this imperative method as an implementation detail.  I think having both might make sense.  The use case I am thinking of, is that some packages (gitlab?) will provide a way to setup the home directory of the user.

**Comment by oxij** [comment]
Actually, this is already kinda supported in NixOS since 8f50d803ef9c94fb82909e22b603982a0a522aea.

in configuration.nix:
~~~~
environment.etc = [
  { source = ./software/config/vimrc;
    target = "skel/.vimrc"; }
  { source = ./software/config/zshrc;
    target = "skel/.zshrc"; }
}

security.pam.services.login.makeHomeDir = true;

users.extraUsers."me" = {
  uid = 1000;
  group = "mygroup";
  useDefaultShell = true;
  home = "/home/me";
  createHome = false; # !!!! this is the crux
};
~~~~

And the home dir will be created the first time you login.

**Comment by jagajaga** [comment]
But how we wll get dotfiles? ln them from /etc?

2015-08-20 18:01 GMT+03:00 Jan Malakhovski <notifications@github.com>:

> Actually, this is already kinda supported in NixOS since 8f50d80
> <https://github.com/NixOS/nixpkgs/commit/8f50d803ef9c94fb82909e22b603982a0a522aea>
> .
>
> in configuration.nix:
>
> environment.etc = [
>   { source = ./software/config/vimrc;
>     target = "skel/.vimrc"; }
>   { source = ./software/config/zshrc;
>     target = "skel/.zshrc"; }
> }
>
> security.pam.services.login.makeHomeDir = true;
>
> users.extraUsers."me" = {
>   uid = 1000;
>   group = "mygroup";
>   useDefaultShell = true;
>   home = "/home/me";
>   createHome = false; # !!!! this is the crux
> };
>
> And the home dir will be created the first time you login.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/9073#issuecomment-133042024>.
>



-- 
Sincerely,
Arseniy Seroka

**Comment by eskimor** [comment]
The pam approach with /etc/skel isn't quite flexible, as all users will get the same files. For the users.<user>.dotFiles approach, I think this would be quite cumbersome and also not very flexible. For example in my setup I have the script clone a private git repo into my home directory. How would you do this decoratively? I made a mistake calling it declarative in the documentation, it is not, but I believe this imperative way of setting up an initial home directory is very worthwhile and useful. Users typically don't manage their home directories in a declarative fashion.

It also imposes no overhead for already created home directories, meaning it does not affect nixos-rebuild's performance at all.

It is tempting to use a declarative approach for managing dot files, but on the other hand who wants to run nixos-rebuild switch every time you make a change to your .emacs?

**Comment by oxij** [comment]
> But how we wll get dotfiles? ln them from /etc?

Yes.

**Comment by oxij** [comment]
> The pam approach with /etc/skel isn't quite flexible, as all users
> will get the same files. ...

Quite a while ago when I thought about implementing something like this
my idea was to have something like the following in `configuration.nix`:

  users.extraUsers.username.dotfiles = {
    ".zshrc".text = ...;
    ".vimrc".text = ...;
  }

which would hash all these the files and generate a login script which
would store these hashes in `~/.nix-dotfiles` like this:

  <hash> .zshrc
  <hash> .vimrc
  ...

or maybe

  <hash> </nix/store/path> .zshrc
  <hash> </nix/store/path> .vimrc
  ...

(for diffs, see below)

and update dotfiles with an init script using the following rules:

* there's a new version of `.nix-dotfiles` available, (*) all the files
  listed in the old `.nix-dotfiles` did not change (their hashes are the
  same as recorded in the old `.nix-dotfiles`) -- write `.nix-dotfiles`
  and update all the things that changed hashes,
* there's a new version of `.nix-dotfiles` available, (*) but some files
  have been changed by the user (their hashes differ from what is
  recorded in the old `.nix-dotfiles`) -- print a warning with all the
  changes (e.g. diffs) and do nothing (this is the only correct way to
  do it, since updates have to be atomic to work in the most general
  case),
* there's a new version of `.nix-dotfiles` available, but there's no old
  `.nix-dotfiles` in user's home -- write `.nix-dotfiles` and apply one
  of the two rules above, reading them from the (*),
* else -- do nothing.

This has all the advantages of plain text dotfiles (e.g. you can hack
them) with the power of nix on top.

I got lazy.

**Comment by eskimor** [comment]
@oxij So basically a system where dot files are overridden by system ones, only if they don't exist or if they haven't changed from previously provided ones? What are the benefits to just managing your dot files in a git repo?

**Comment by oxij** [comment]
Yes, atomically.

Automatic deploy. And you can generate a dotfile using a nix function
that depends on user's login name, home dir or whatever, for example.

**Comment by oxij** [comment]
Ah, I did not describe the algorithm correctly, yes.
The typical workflow would be:

1. Hack dotfiles around till they work as expected
2. Run system update (or some diff-like util) to see the diff
2. Update the generator with that diff
3. Run system update to check that you applied it correctly (no warnings)

The point is that if you made correct changes to the dotfiles generator so that it would generate exactly the dotfiles you already have in your home, it would assume you have already made the updates and it's just the `.nix-dotfiles` that needs updating.

**Comment by oxij** [comment]
In the simplest case when your dotfiles just symlink to files in your dotfiles git repo (or maybe the are the files in your git repo, like when using `vcsh`)
and all the `configuration.nix` does is lists how to deploy them, you just update them in place in your home dir and the system update does exactly nothing or prints a warning for your user while updating other users, that use same dotfiles.

**Comment by eskimor** [comment]
I am sorry, but in comparison to a git repo I still only see a more complex system without any real benefits. The home directory is impure, so reproducibility is a bit limited. For git it requires that you checked everything in, in their latest versions. For your approach it is pretty much the same, but with different tools/workflows. Am I still missing something?

**Comment by oxij** [comment]
Okay, but what the benefit of your patch then?

**Comment by eskimor** [comment]
Well it is just a minor convenience that you have your system fully set up, including your home directory with one command: "nixos-install". That's it, nothing more. I think it is nice and it costs nothing. 

It is kinda cool to have your complete system set up with just one command.

**Comment by oxij** [comment]
But then, if you want updates, you have to have some update logic. If
you don't want updates, why automate something that is run exactly once.

I'd say `vcsh` (possibly with `mr`) is a much better solution than any
of these here if you don't want to functionally generate configuration
files.

**Comment by eskimor** [comment]
It is run exactly once for each system & user you have. The script that is run can be git clones, mr commands whatever. So it is not his patch or git/mr, but this patch together with git/mr.

It is really just a minor feature, because you could of course run nixos-install  and afterwards a script which sets up home directories. But with this patch you have everything in one place, which is in the spirit of nixos, I think. For example if you want to remove a user for your setups, you do this in one place instead of in two. If you add a new user to your configuration and have the git repo on your server already, then all you need to to is add the user to your configuraion.nix file and nix takes care of everything.

- One command installation
- Integrated user management includes setting up home directories


Whether you find this useful or not depends on your workflow. But what is nice - not even does it cost you nothing if you don't use it, it even costs you nothing if you use it. Because if the home directories are already present ( a check that is run already) nothing happens.

Also this patch it is not necessarily your system or this patch, there could easily be both and people choose what they prefer.

**Comment by eskimor** [comment]
Closing, due to lack of interest.

**Pull request closed without merging**